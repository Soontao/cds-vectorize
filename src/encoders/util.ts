import { Vector } from "./type.js";

export function isNull(value: any): boolean {
  return value === null || value === undefined || (typeof value == "number" && isNaN(value));
}

export function nullVector(length: number): Vector {
  return new Array(length).fill(-(2 ** 16));
}
