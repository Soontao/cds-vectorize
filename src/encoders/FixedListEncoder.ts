import { Vector } from "../type";
import Encoder from "./Encoder";
import ObjectMetadata from "./Metadata";
import ObjectEncoder, { sortMetaAndFillEncoders } from "./ObjectEncoder";

export function match<T = any>(obj: T, part: any): part is Partial<T> {
  return Object.entries(part).every(([p, v]) => (obj as any)[p] === v);
}

/**
 * @human (+)
 * @ai (-)
 */
export class FixedListEncoder<T = any> implements Encoder<Array<T>> {
  #meta: ObjectMetadata;

  #objectEncoder: ObjectEncoder<T>;

  #length: number;

  #positionDict: Partial<T>[];

  /**
   *
   * @param meta metadata of item of list
   * @param fixedLength fixed length of list, means the max items
   * @param positionDict
   */
  constructor(meta: ObjectMetadata, positionDict: Array<Partial<T>>) {
    this.#meta = sortMetaAndFillEncoders(meta);
    this.#objectEncoder = new ObjectEncoder(this.#meta);
    this.#length = positionDict.length * this.#objectEncoder.length;
    this.#positionDict = positionDict;
  }

  features(name: string = "root"): string[] {
    return this.#positionDict.map((_, idx) => this.#objectEncoder.features(`${name}_${idx}`)).flat();
  }

  encode(value: T[]): Vector {
    const encodedVector: number[] = new Array(0);

    for (const partialItem of this.#positionDict) {
      const item = value.find((obj) => match(obj, partialItem));
      if (!item) {
        encodedVector.push(...new Array(this.#objectEncoder.length).fill(NaN));
      } else {
        const itemVector = this.#objectEncoder.encode(item!);
        encodedVector.push(...itemVector);
      }
    }

    return encodedVector;
  }

  decode(vec: Vector): T[] {
    if (vec.length !== this.length) {
      throw new Error("FixedListEncoder: Invalid vector length");
    }

    const decodedList: T[] = [];

    for (let i = 0; i < this.#positionDict.length; i++) {
      const startIndex = i * this.#objectEncoder.length;
      const endIndex = startIndex + this.#objectEncoder.length;

      const itemVector = vec.slice(startIndex, endIndex);

      // Check if all values in the itemVector are NaN
      if (itemVector.every((val) => isNaN(val))) {
        // do nothing
      } else {
        const decodedItem = this.#objectEncoder.decode(itemVector);
        decodedList.push(decodedItem);
      }
    }

    return decodedList;
  }

  get length(): number {
    return this.#length;
  }
}
