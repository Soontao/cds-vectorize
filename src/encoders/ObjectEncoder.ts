/* eslint-disable camelcase */
import { inspect } from "util";
import BoolEncoder from "./BoolEncoder.js";
import CategoryEncoder from "./CategoryEncoder.js";
import DateTimeEncoder from "./DateTimeEncoder.js";
import Encoder from "./Encoder.js";
import FixedListEncoder from "./FixedListEncoder.js";
import ObjectMetadata, { metadataValidator } from "./Metadata.js";
import NumericEncoder from "./NumericEncoder.js";
import StatisticListEncoder from "./StatisticListEncoder.js";
import UUIDEncoder from "./UUIDEncoder.js";
import { Vector } from "./type.js";

export function sort(meta: ObjectMetadata): ObjectMetadata {
  if (meta._sorted) {
    return meta;
  }
  const newMeta = { ...meta };
  newMeta.properties = meta.properties
    .map((v) => {
      if (v.type == "object") {
        const newV = { ...v };
        newV.meta = sort(v.meta as any);
        return newV;
      }
      if (v.type === "category" && v.values) {
        v.values = v.values!.sort();
      }
      return v;
    })
    .sort((p1, p2) => (p1.name > p2.name ? 1 : -1));
  Object.defineProperty(newMeta, "_sorted", { value: true, enumerable: false });
  return newMeta;
}

export function fillEncoders(meta: ObjectMetadata) {
  if (meta._encoder_filled) {
    return meta;
  }
  for (const property of meta.properties) {
    if (property._encoder !== undefined) {
      continue;
    }
    if (property.meta) {
      sortMetaAndFillEncoders(property.meta!); // inner meta create encoders firstly
    }
    switch (property.type) {
      case "bool":
        property._encoder = new BoolEncoder();
        break;
      case "category":
        property._encoder = new CategoryEncoder(property.values as Array<any>);
        break;
      case "datetime":
        property._encoder = new DateTimeEncoder();
        break;
      case "numeric":
        property._encoder = new NumericEncoder();
        break;
      case "object":
        property._encoder = new ObjectEncoder(property.meta!);
        break;
      case "uuid":
        property._encoder = new UUIDEncoder();
        break;
      case "fixed_object_list":
        property._encoder = new FixedListEncoder(property.meta!, property.position_dict!);
        break;
      case "statistic_object_list":
        property._encoder = new StatisticListEncoder(property.meta!);
        break;
      default:
        throw new TypeError(`cannot handle type ${property.type} for ${property.name}`);
    }
  }
  calculateObjectVecLength(meta);
  Object.defineProperty(meta, "_encoder_filled", { value: true, enumerable: false });
  return meta;
}

export function sortMetaAndFillEncoders(meta: ObjectMetadata): ObjectMetadata {
  if (meta._valid == undefined) {
    const isValid = metadataValidator(meta as any);
    if (!isValid) {
      throw new Error(`not valid metadata, please check ${inspect(metadataValidator.errors)}`);
    }
  }
  Object.defineProperty(meta, "_valid", { value: true, enumerable: false });
  return fillEncoders(sort(meta));
}

export function calculateObjectVecLength(meta: ObjectMetadata): number {
  if (meta._length) {
    return meta._length;
  }
  const _length = meta.properties.reduce((acc, property) => acc + property._encoder!.length, 0);
  Object.defineProperty(meta, "_length", { value: _length, enumerable: false });
  return _length;
}

/**
 * @ai
 * @human
 */
export class ObjectEncoder<T> implements Encoder<T> {
  #meta: ObjectMetadata;

  constructor(meta: ObjectMetadata) {
    this.#meta = sortMetaAndFillEncoders(meta);
  }

  features(name: string = "root"): string[] {
    return this.#meta.properties.map((p) => p._encoder!.features(p.name).map((pName) => `${name}_${pName}`)).flat();
  }

  encode(value: T): Vector {
    if (typeof value !== "object" || value === null) {
      throw new Error("Invalid value. Expected an object.");
    }

    const encodedVector: number[] = [];

    for (const property of this.#meta.properties) {
      encodedVector.push(...property._encoder!.encode((value as any)[property.name]));
    }

    return encodedVector;
  }

  decode(vec: Vector): T {
    if (vec.length !== this.length) {
      throw new Error("Invalid vector length");
    }

    const decodedObject: any = {};
    let index = 0;

    for (const property of this.#meta.properties) {
      const encoder = property._encoder!;
      const value = encoder.decode(vec.slice(index, index + encoder.length));
      decodedObject[property.name] = value;
      index += property._encoder!.length;
    }

    return decodedObject as T;
  }

  get length(): number {
    return this.#meta._length!;
  }
}

export default ObjectEncoder;
