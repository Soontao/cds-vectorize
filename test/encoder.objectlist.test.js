// @human
// @ai
import { FixedListEncoder } from "../src/encoders/FixedListEncoder.js";
import { ObjectEncoder } from "../src/encoders/ObjectEncoder.js";
import { nullVector } from "../src/encoders/util.js";

describe("FixedListEncoder", () => {
  const meta = {
    properties: [
      { name: "name", type: "category", values: ["Alice", "Bob", "Charlie"] },
      { name: "age", type: "numeric" },
    ],
  };

  const positionDict = [
    { name: "Alice", age: 25 },
    { name: "Bob", age: 30 },
  ];

  const encoder = new FixedListEncoder(meta, positionDict);

  const testData = [
    { name: "Alice", age: 25 },
    { name: "Bob", age: 30 },
    { name: "Charlie", age: 35 },
  ];

  it("should encode and decode correctly", () => {
    const encoded = encoder.encode(testData);
    expect(encoded).toMatchSnapshot();
    const decoded = encoder.decode(encoded);
    expect(decoded).toHaveLength(2);
    expect(decoded).toMatchSnapshot();
  });

  it("should encode with NaN for missing items", () => {
    const objEncoder = new ObjectEncoder(meta);
    const testDataSubset = [
      { name: "Alice", age: 25 },
      { name: "Charlie", age: 35 },
    ];

    const encoded = encoder.encode(testDataSubset);

    // Check if NaN values are present for the missing item
    expect(encoded).toEqual([
      // Encoded Alice
      ...objEncoder.encode({ name: "Alice", age: 25, length: 2 }),
      // Encoded Bob (missing)
      ...nullVector(objEncoder.length),
    ]);
  });

  it("should handle invalid vector length during decoding", () => {
    const invalidVector = [1, 2, 3, 4, 5];

    expect(() => encoder.decode(invalidVector)).toThrow("FixedListEncoder: Invalid vector length");
  });

  it("should handle decoding with NaN list", () => {
    const nanEncoded = nullVector(8);

    const decoded = encoder.decode(nanEncoded);

    // Check if the decoded list contains undefined for each item
    expect(decoded).toHaveLength(0);
  });

  it("should handle encoding and decoding an empty input list", () => {
    const emptyInputList = [];

    const encoded = encoder.encode(emptyInputList);
    const decoded = encoder.decode(encoded);

    // Check if the decoded list is empty
    expect(decoded).toHaveLength(0);
  });

  it("should handle encoding and decoding an empty input list", () => {
    const emptyInputList = [null, undefined, {}];

    const encoded = encoder.encode(emptyInputList);
    const decoded = encoder.decode(encoded);

    expect(decoded).toHaveLength(0);
  });

  it("should encode and decode correctly with nested object in list", () => {
    const meta = {
      properties: [
        { name: "name", type: "category", values: ["Alice", "Bob", "Charlie"] },
        { name: "age", type: "numeric" },
        {
          name: "details",
          type: "object",
          meta: {
            properties: [
              { name: "gender", type: "category", values: ["Male", "Female"] },
              { name: "city", type: "category", values: ["New York", "London"] },
            ],
          },
        },
      ],
    };

    const positionDict = [{ name: "Alice" }, { name: "Bob" }];

    const encoder = new FixedListEncoder(meta, positionDict);

    const testData = [
      { name: "Alice", age: 25, details: { gender: "Female", city: "New York" } },
      { name: "Bob", age: 30, details: { gender: "Male", city: "London" } },
      { name: "Charlie", age: 35, details: { gender: "Female", city: "Paris" } },
    ];
    const encoded = encoder.encode(testData);
    const features = encoder.features();
    expect(features).toMatchSnapshot();
    expect(encoded).toMatchSnapshot();
    expect(encoded).toHaveLength(features.length);

    const decoded = encoder.decode(encoded);
    expect(decoded).toHaveLength(2);
    expect(decoded).toMatchSnapshot();
  });
});
