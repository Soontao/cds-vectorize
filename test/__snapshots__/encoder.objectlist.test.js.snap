// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixedListEncoder should encode and decode correctly 1`] = `
[
  25,
  1,
  0,
  0,
  30,
  0,
  1,
  0,
]
`;

exports[`FixedListEncoder should encode and decode correctly 2`] = `
[
  {
    "age": 25,
    "name": "Alice",
  },
  {
    "age": 30,
    "name": "Bob",
  },
]
`;

exports[`FixedListEncoder should encode and decode correctly with nested object in list 1`] = `
[
  "root_0_age",
  "root_0_details_city_is_London",
  "root_0_details_city_is_New York",
  "root_0_details_gender_is_Female",
  "root_0_details_gender_is_Male",
  "root_0_name_is_Alice",
  "root_0_name_is_Bob",
  "root_0_name_is_Charlie",
  "root_1_age",
  "root_1_details_city_is_London",
  "root_1_details_city_is_New York",
  "root_1_details_gender_is_Female",
  "root_1_details_gender_is_Male",
  "root_1_name_is_Alice",
  "root_1_name_is_Bob",
  "root_1_name_is_Charlie",
]
`;

exports[`FixedListEncoder should encode and decode correctly with nested object in list 2`] = `
[
  25,
  0,
  1,
  1,
  0,
  1,
  0,
  0,
  30,
  1,
  0,
  0,
  1,
  0,
  1,
  0,
]
`;

exports[`FixedListEncoder should encode and decode correctly with nested object in list 3`] = `
[
  {
    "age": 25,
    "details": {
      "city": "New York",
      "gender": "Female",
    },
    "name": "Alice",
  },
  {
    "age": 30,
    "details": {
      "city": "London",
      "gender": "Male",
    },
    "name": "Bob",
  },
]
`;
